### PR TITLE
[Woo POS] Add toolbar to the cart's subscreen

### LINKED ISSUE #11762: [Woo POS] Add "toolbar" to the cart subscreen
Image

Image

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #11762

### Description

Adds toolbar actions to the cart subsceens
image

### Testing information

Try adding/removing products. Remove all products. Navigate back using newly added back arrow

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4923871/02a91930-f595-4dbb-9510-36ce942b763b
image

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>WooPosCartToolbar</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt [hunk 1]
```diff
@@ -22,19 +26,25 @@ class WooPosCartViewModel @Inject constructor(
...
 ) : ViewModel() {
     private val _state = savedState.getStateFlow(
         scope = viewModelScope,
         initialValue = WooPosCartState(),
         key = "cartViewState"
     )
-    val state: StateFlow<WooPosCartState> = _state
+
+    val state: LiveData<WooPosCartState> = _state
+        .asLiveData()
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:39]: 🤔 Why LiveData?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt [hunk 2]
```diff
@@ -16,3 +20,14 @@ data class WooPosCartListItem(
...
+
+@Parcelize
+data class WooPosToolbar(
+    @DrawableRes val icon: Int = R.drawable.ic_shopping_cart,
+    val itemsCount: String = "",
+    val isClearAllButtonVisible: Boolean = false,
+) : Parcelable
+
+enum class WooPosCartStatus {
+    IN_PROGRESS, CHECKOUT
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt]: 💡 `IN_PROGRESS` sounds a bit like a loading state. How about `EDITABLE` or `ORDER_CREATION`?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt [hunk 3]
```diff
@@ -16,3 +20,14 @@ data class WooPosCartListItem(
     val productId: Long,
     val title: String,
 ) : Parcelable
+
+@Parcelize
+data class WooPosToolbar(
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt]: 💡 n.p. I'd consider a less generic `WooPosCartToolbar`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartScreen.kt [hunk 4]
```diff
@@ -94,6 +101,58 @@ private fun WooPosCartScreen(
...
+    onBackClicked: () -> Unit
+) {
+    Row(
+        modifier = Modifier.fillMaxWidth(),
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        IconButton(onClick = { onBackClicked() }) {
+            Icon(
+                imageVector = ImageVector.vectorResource(toolbar.icon),
+                contentDescription = "cart icon",
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartScreen.kt]: 💡 Should be extracted

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt [hunk 5]
```diff
@@ -48,15 +58,31 @@ class WooPosCartViewModel @Inject constructor(
...
+
+                sendEventToParent(ChildToParentEvent.BackFromCheckoutToCartClicked)
+            }
+
+            WooPosCartUIEvent.ClearAllClicked -> {
+                val currentState = _state.value
+                if (currentState.isOrderCreationInProgress) return
+
+                _state.value = currentState.copy(
+                    itemsInCart = emptyList()
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:77]: ❓ Shouldn't we reset the state fully, e.g. disable the checkout button?

**Review by samiuelson — APPROVED** [review]
Works well! I added a few minor suggestions, please take a look.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:39]: I could test that with "map"  🥺. I'll try again

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt]: `EDITABLE` sounds good. Changed it here - 320b8c19868ad0922fe0e90c751ef46ea2a7168a

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt (hunk 3 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartState.kt]: Missed that - 5a5a67a2974d21d91a5f3f118f4e52ad3c37efa9

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartScreen.kt (hunk 4 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartScreen.kt]: Done a4f7d2d4847556fcf47ffb65241d940af94b1521

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:39]: Actually, not sure if it's worth it. With map and Stateflow you have to do something like that:

```
    val state: StateFlow<WooPosCartState> = _state
        .map { updateToolbarState(it) }
        .map { updateStateDependingOnCartStatus(it) }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(),
            initialValue = WooPosCartState()
        )
```

Which is weird as the initial value is set in 2 places. Also testing wise I didn't find how to subscribe to it without blocking and without starting a new job which looks weird and overcomplicated

I may do something wrong - open to suggestions and will update this in the following PRs

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt (hunk 5 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt:77]: That;s handled in `updateStateDependingOnCartStatus` - it updates visibilities based on the cart status

**Pull request merged**